Add new blog "Peeling the Big Onion" (#1923)
adoptium/adoptium.net: content/blog/peeling-the-big-onion/index.md

---
title: Peeling the Big Onion - Stripping out layers of indirection from test frameworks
date: "2023-06-22T11:00:00+00:00"

> author: Mesbah-Alam

Reviewer: ```suggestion
author: mesbahalam
```

The author should be set as the key in the JSON file

description: This article demonstrates the need to remove layers of redirection from QA automation frameworks.
tags:
  - Adoptium
  - Test
  - QA Automation
---

## Background

> Automation test frameworks should run test in the simplest way possible. However, we often over-engineer them by attempting to make one grand solution to support automation of a wide variety of tests. This results in layers of indirection, unnecessary technical debt, and typically— inefficiency. “Peeling the onion”, so to speak, thus becomes an imperative exercise to ensure a test framework's evolution in the right direction.

Reviewer 2: ```suggestion
Automation test frameworks should run tests in the simplest way possible. However, we often over-engineer them by attempting to make one grand solution to support automation of a wide variety of tests. This results in layers of indirection, unnecessary technical debt, and typically— inefficiency. “Peeling the onion”, so to speak, thus becomes an imperative exercise to ensure a test framework's evolution in the right direction.
```

> Automation test frameworks should run tests in the simplest way possible. However, we often over-engineer them by attempting to make one grand solution to support automation of a wide variety of tests. This results in layers of indirection, unnecessary technical debt, and typically— inefficiency. “Peeling the onion”, so to speak, thus becomes an imperative exercise to ensure a test framework's evolution in the right direction.

Author: Reading the above quoted paragraph again, It makes sense and I am totally okay with it. Accepted all changes.

> Automation test frameworks should run tests in the simplest way possible. However, we often over-engineer them by attempting to make one grand solution to support automation of a wide variety of tests. This results in layers of indirection, unnecessary technical debt, and typically— inefficiency. “Peeling the onion”, so to speak, thus becomes an imperative exercise to ensure a test framework's evolution in the right direction.

Reviewer 3: Perhaps it could be rephrased:
This is the case with some of the automation scripts in the [AQAvit project](https://projects.eclipse.org/projects/adoptium.aqavit). The System Test Framework ([STF](https://github.com/adoptium/STF)) and [aqa-systemtest](https://github.com/adoptium/aqa-systemtest) which was originally contributed by IBM contain some of the automation that was also initially used for triggering the runs of the licensed Java Technology Compatibility Tests (TCKs) as part of the Temurin Compliance project.

## The Big onion

> Let’s use a real-world example. We run a large set of TCKs as part of our Java compatibility verification process of Semeru OpenJ9 and IBM JDKs. Historically, these TCKs used to be driven via an age-old cryptic Perl wrapper which was difficult to maintain and evolve to fit the needs of newer JDK versions.

Reviewer 2: ```suggestion
Let’s use a real-world example. We run a large set of Java Technology Compatibility Tests (TCKs) as part of our Java compatibility verification process of Semeru OpenJ9 and IBM JDKs. Historically, these TCKs used to be driven via an age-old cryptic Perl wrapper which was difficult to maintain and evolve to fit the needs of newer JDK versions.
```

Who is "we" in this sentence? Not Adoptium.

> At one point we decided to replace this legacy TCK Perl wrapper with a new test framework that was being designed to drive stress tests— [STF](https://github.com/adoptium/STF). However, STF was simply too lofty for the needs of TCK execution. First, STF would need users to define test cases as a series of conceptual 'stages' using Java. It would then take those ‘stages’ and generate a set of Perl scripts to orchestrate the second layer of Java command-lines which would culminate in the eventual automated test job in the CI system. While this has benefits in terms of debug-ability of load/stress tests, it was not the best way to drive TCK, that itself comes with its own sophisticated test harness.

Reviewer 2: ```suggestion
At one point we decided to replace this legacy TCK Perl wrapper with a new test framework that was being designed to drive stress tests - the [System Test Framework (STF)](https://github.com/adoptium/STF). However, STF was simply too lofty for the needs of TCK execution. First, STF would need users to define test cases as a series of conceptual 'stages' written in Java code. It would then take those ‘stages’ and generate a set of Perl scripts to orchestrate the second layer of Java command-lines which would culminate in the eventual automated test job in the Continuous Integration (CI) system. While this has benefits in terms of debug-ability of load/stress tests, it was not the best way to drive the TCK, that itself comes with its own sophisticated test harness.
```

> <center><img src="bigo.jpg" alt="The Big Onion" width="250"/></center>

Reviewer: you shouldn't need this html syntax as gatsby will do a good enough job rendering the images, please stick to markdown syntax

Author: According to the "preview" on my Eclipse, markdown was rendering the image too large and also not centre-aligning. If you know how to specify those settings in markdown sysntax, let me know.

Reviewer: Ignore the preview you see, the website has a load of rendering that optimises the images

> Result—the ‘Big Onion’. By adopting STF to run TCKs, not only did we introduce unnecessary layers of indirection in our automation story `(Java -> Perl -> Java)`, but we also ended up incurring additional compilation time in TCK builds in our CI system, since STF requires checking out and compiling of both its own Git repository, as well as one extra Git repository that of the System tests.

Reviewer 2: ```suggestion
Result—the ‘Big Onion’. By adopting STF to run TCKs, not only did we introduce unnecessary layers of indirection in our automation story `(Java -> Perl -> Java)`, but we also ended up incurring additional compilation time in TCK builds in our CI system, since STF requires checking out and compiling from both its own Git repository, as well as one extra Git repository that of the System tests.
```

## Peel the onion

> So, to run the TCK harness with less indirection and more efficiency, we decided to decouple it from the STF stress test framework altogether and replace it with [JavaTestRunner](https://github.com/adoptium/aqa-tests/blob/master/jck/jtrunner/JavaTestRunner.java)– a simple Java class that can perform the essential tasks of generating command files for TCK, building result summary, starting required services and the managing the execution TCK harness itself.

Reviewer 2: ```suggestion
So, to run the TCK harness with less indirection and more efficiency, we decided to decouple it from the STF stress test framework altogether and replace it with [JavaTestRunner](https://github.com/adoptium/aqa-tests/blob/master/jck/jtrunner/JavaTestRunner.java)– a simple Java class that can perform the essential tasks of generating command files for TCK, building a result summary, starting required services and managing execution of the TCK harness itself.
```

![The Peeled Onion](bigo-peeled.jpg)

> Unlike the STF framework, JavaTestRunner is housed within the same Adoptium aqa-tests repository, like the rest of the AQAvit test suites. This relieved us from having to checkout and compile two additional repositories— resulting in a dramatic reduction of compile time for TCK CI jobs (we ended up saving ~3285 hours of machine time per year).

Reviewer 2: ```suggestion
Unlike the STF framework, JavaTestRunner is housed within the same [Adoptium aqa-tests repository](https://github.com/adoptium/aqa-tests) as the rest of the AQAvit test suites. This relieved us from having to checkout and compile two additional repositories— resulting in a dramatic reduction of compile time for TCK CI jobs (we ended up saving ~3,285 hours of machine time per year).
```

## Peel it more

> Stripping out layers of indirection from test frameworks typically happens in iterations. JavaTestRunner gave us the ability to run TCK jobs faster, with more debug-ability and less indirection. However, it still was an ‘onion’. Meaning, few more layers could potentially be peeled off still.

Reviewer 2: ```suggestion
Stripping out layers of indirection from test frameworks typically happens in iterations. JavaTestRunner gave us the ability to run TCK jobs faster, with more debug-ability and less indirection. However, it still was an ‘onion’. Meaning a few more layers could potentially be peeled off still.
```

> One such layer was how JavaTestRunner drives the TCK harness itself. TCK harness is a Java application itself. JavaTestRunner would gather all the inputs from user via playlist, construct the command-line and run the TCK harness via ProcessBuilder. This layer of indirection `(JVM->JVM)` is not only inefficient, but it also results in hiding test command-lines— making the TCK builds harder to triage.

Reviewer 2: ```suggestion
One such layer was how JavaTestRunner drives the TCK harness. The TCK harness is a Java application. JavaTestRunner would gather all the inputs from the user via a playlist, construct the command-line and run the TCK harness via ProcessBuilder. This layer of indirection `(JVM->JVM)` is not only inefficient, but it also results in hiding test command-lines— making the TCK builds harder to debug.
```

> The above issue was solved by stripping out JavaTestRunner and replacing it with [JavaTestUtil](https://github.com/adoptium/aqa-tests/blob/master/jck/jtrunner/JavatestUtil.java). JavaTestUtil performs the essential tasks of command file generation and result summary generation only, while the actual command-lines to start required services and running the TCK harness itself are placed in a makefile outside of the Java wrapper. This means one less layer of indirection and faster execution. Plus, users can now have more visibility to the actual command-lines that get executed for each TCK test target.

Reviewer 2: ```suggestion
The above issue was solved by stripping out JavaTestRunner and replacing it with [JavaTestUtil](https://github.com/adoptium/aqa-tests/blob/master/jck/jtrunner/JavatestUtil.java). JavaTestUtil performs the essential tasks of command file generation and result summary generation only, while the actual command-lines to start required services and running the TCK harness are placed in a makefile outside of the Java wrapper. This means one less layer of indirection and faster execution. Plus, users now have more visibility to the actual command-lines that are executed for each TCK test target.
```

## Summary

> Test frameworks evolve into a 'Big Onions'. Peeling those onions should be considered among software development best practices. It helps development teams to reduce technical debt while continuously improving QA automation efficiency.

Reviewer 2: ```suggestion
Test frameworks evolve into a 'Big Onions'. Peeling those onions should be considered among software development best practices. It helps development teams to reduce technical debt while continuously improving quality assurance automation efficiency.
```

---
title: Peeling the Big Onion - Stripping out layers of indirection from test frameworks
date: "2023-06-22T11:00:00+00:00"
author: mesbahalam
description: This article demonstrates the need to remove layers of redirection from QA automation frameworks.
tags:
  - Adoptium
  - Test
  - QA Automation
---

## Background

Automation test frameworks should run tests in the simplest way possible. However, we often over-engineer them by attempting to make one grand solution to support automation of a wide variety of tests. This results in layers of indirection, unnecessary technical debt, and typically— inefficiency. “Peeling the onion”, so to speak, thus becomes an imperative exercise to ensure a test framework's evolution in the right direction.

> As with many large open-source efforts, some automation and tooling come via contributions from participating members of the community. This is the case with some of the automation scripts in the [AQAvit project](https://projects.eclipse.org/projects/adoptium.aqavit). The System Test Framework ((STF)[https://github.com/adoptium/STF]), [aqa-systemtest](https://github.com/adoptium/aqa-systemtest) and thus some of the automation that is also used for triggering the runs of the licensed Java Technology Compatibility Tests (TCKs) was contributed by IBM.

Reviewer 3: ```suggestion
As with many large open-source efforts, some automation and tooling come via contributions from participating members of the community. This is the case with some of the automation scripts in the [AQAvit project](https://projects.eclipse.org/projects/adoptium.aqavit). The System Test Framework ([STF](https://github.com/adoptium/STF)) and [aqa-systemtest](https://github.com/adoptium/aqa-systemtest) which was originally contributed by IBM contain some of the automation that was also initially used for triggering the runs of the licensed Java Technology Compatibility Tests (TCKs) as part of the Temurin Compliance project.
```

It is important to note the difference between automation used to trigger "any" test material versus the TCK test material itself and the rules that govern that it be kept private. The TCKs are run in private by distributors with a license to that material. At Adoptium, this activity is done in the [Temurin Compliance project](https://projects.eclipse.org/projects/adoptium.temurin-compliance).

## The Big onion

> This is a story of reducing complexity. Given the provenance of some of the automation scripts, let’s look at a real-world example of the simplification of test pipelines. A large set of Java Technology Compatibility Tests (TCKs) is run as part of Java compatibility verification process of all compliant JDK distributions, including Eclipse Temurin binaries, Semeru OpenJ9 certified edition, IBM JDKs and any distribution being listed in the [Adoptium marketplace](https://adoptium.net/marketplace/). Historically for IBM JDKs, these TCKs used to be driven via an age-old cryptic Perl wrapper which was difficult to maintain and evolve to fit the needs of newer JDK versions.

Reviewer: ```suggestion
This is a story of reducing complexity. Given the provenance of some of the automation scripts, let’s look at a real-world example of the simplification of test pipelines. A large set of Java Technology Compatibility Tests (TCKs) is run as part of Java compatibility verification process of all compliant JDK distributions, including Eclipse Temurin binaries, Semeru OpenJ9 certified edition, IBM JDKs and any distribution being listed in the [Adoptium marketplace](/marketplace/). Historically for IBM JDKs, these TCKs used to be driven via an age-old cryptic Perl wrapper which was difficult to maintain and evolve to fit the needs of newer JDK versions.
```
